### PR TITLE
feature: Pub Edge Proposal API Endpoint

### PR DESCRIPTION
**Pull request by 3mcd**
This PR adds a `/pubEdgeProposal` endpoint that is used to create a partial PubEdge when given a Pub URL or DOI. The endpoint is currently missing support for doi.org URLs and external resources.

Example requests:

- Pub: http://localhost:9876/api/pubEdgeProposal?object=https%3A%2F%2Fjods.duqduq.org%2Fpub%2F5qv35sci%2Frelease%2F1
- DOI for Pub: http://localhost:9876/api/pubEdgeProposal?object=10.21428%2F7808da6b.00ff17d0
- DOI: http://localhost:9876/api/pubEdgeProposal?object=10.1155%2F2014%2F413629

Partial #883

### CODE REVIEW

File: server/pubEdgeProposal/api.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+			if (!pub) {
+				// URL is not PubPub
+				// TODO: Scrape target for DOI
+				return res.status(404);
+			}
+
+			edge = createPubEdgeProposalFromPub(pub, url);
+		} else if (isDoi(object)) {
+			const response = await fetch(`https://api.crossref.org/works/${object}`);
```
Comment (idreyn) [inline server/pubEdgeProposal/api.js:63]: As a matter of taste, I would factor the entire process of fetching from crossref and creating the edge proposal object into `./queries`.

File: server/pubEdgeProposal/api.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import { isDoi } from 'utils/dois';
+
+import { getPubDataFromUrl } from './queries';
+
+const createPubEdgeProposalFromPub = (pub, url) => {
+	const { attributions, avatar, description, doi, lastPublishedAt, title } = pub;
+	const contributors = attributions.map((a) => a.name);
+
+	return {
+		targetPub: {
```
Comment (idreyn) [inline server/pubEdgeProposal/api.js:14]: ⚠️ Don't cast `Pub` to the `ExternalPublication` shape — for the sake of consistency with the rest of the `<PubEdge>` callsites, we should provide a Pub with the associated models required here. Those turn out to be specified [by the `getOptionsForIncludedPub` helper function in my PR](https://github.com/pubpub/pubpub/blob/cc7507cfd561e5dcabcb9fbb16242bf4d6d1fa1f/server/utils/queryHelpers/edgeOptions.js#L10).

You could choose to copy this code, interline it, or wait for my PR to land to resolve this.

File: server/pubEdgeProposal/api.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+import app, { wrap } from 'server/server';
+import { parseUrl } from 'utils/urls';
+import { isDoi } from 'utils/dois';
+
+import { getPubDataFromUrl } from './queries';
+
+const createPubEdgeProposalFromPub = (pub, url) => {
+	const { attributions, avatar, description, doi, lastPublishedAt, title } = pub;
+	const contributors = attributions.map((a) => a.name);
```
Comment (idreyn) [inline server/pubEdgeProposal/api.js:11]: (The comment below explains why this line is actually not needed, so the following is just FYI)

Not every `PubAttribution` object has a `name`. Those linked to a PubPub user will get their name info from the user model instead. Customarily we map attributions over `ensureUserForAttribution`, which will append a dummy user object to the attribution if one doesn't already exist, and then to look at `attribution.user.fullName`. Sounds a bit weird, I know — I am super open to a more robust tool for solving this problem!

File: server/pubEdgeProposal/api.js (hunk 2 above)
Comment (3mcd) [reply server/pubEdgeProposal/api.js:14]: I'll wait for your PR.

File: package.json [hunk 4]
```diff
@@ -124,6 +124,7 @@
 		"mutationobserver-shim": "^0.3.3",
 		"newrelic": "^5.7.0",
 		"no-slash": "^1.2.15",
+		"node-fetch": "^2.6.0",
```
[the remark this answers is not in the document]
Comment (3mcd) [reply package.json:127]: I'll use `request-promise` for the sake of consistency.

File: client/components/PubEdgeListing/PubEdgeListing.js [hunk 5]
```diff
@@ -48,8 +48,20 @@ const PubEdgeListing = (props) => {
 	const [filters, setFilters] = useState(initialFilters);
 	const filteredPubEdges = filterPubEdges(filters, pubEdges);
 	const { [index]: active, length } = filteredPubEdges;
-	const next = useCallback(() => setIndex((i) => (i + 1) % length), [length]);
-	const back = useCallback(() => setIndex((i) => (i - 1 + length) % length), [length]);
+	const next = useCallback(
+		(e) => {
```
[the remark this answers is not in the document]
Comment (3mcd) [reply client/components/PubEdgeListing/PubEdgeListing.js]: Nope, #897 fixed it!

**Pull request merged**